## FIFA 2021 DATA ANALYSIS 
#### Dataset: The data was gotten from kaggle and can be found [Here]('https://www.kaggle.com/datasets/yagunnersya/fifa-21-messy-raw-dataset-for-cleaning-exploring')

> *The dataset consist of various players name including their nationality, club, age, height, assist, e.t.c.*

## Importing Necessary Libraries


In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
%matplotlib inline 

### Loading the dataset

In [2]:
data = pd.read_csv('C:/Users/Taiwo_Victoria/Desktop/DATA ANALYSIS PROJECTS/FIFA 21 messy_raw dataset for cleaning exploring/dataset/fifa21_raw_data_v2.csv',low_memory = False )

## Data preparation 

In [3]:
data.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club  ...     A/W     D/W   IR PAC SHO  PAS  \
0   93         \n\n\n\nFC Barcelona  ...  Medium     Low  5 ★  85  92   91   
1   92             \n\n\n\nJuventus  ...    High     Low  5 ★  89  93   81   
2   93      \n\n\n\nAtlético Madrid  ...  Medium  Medium  3 ★  87  92   78   
3   91      \n\n\n\nManchester City  ...    High    High  4 ★  76  86   93   
4   91  \n\n\n\nParis Saint-Germain  ...    High  Medium  5 ★  91  85   86   

  DRI DEF PHY Hits  
0  95  38  65  771  
1  89  35  77  562  
2  90  52  90  150  
3  88  64  78  207  
4  94  36  59  595  

[5 rows x 77 columns]

In [4]:
data.shape

(18979, 77)

> There are 77 columns in the dataset. Since all the columns will not be considered, some of the columns will be extracted out.

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

From this analysis, there are some questions that will be answered from the dataset.

Questions:

*1.* Who is the oldest fifa21 player?

*2.* Who are the 10 most paid players? 

*3.* Which players have been playing at a club for more than 10 years.

*4.* Which players are highly valuable but still underpaid(on low wages).

*5.* What is the highest player loan period?

*6.* Who is the tallest player?

*7.* Which player as the longest conntract with his club?

*8.* Who are the best players based on their positioning?

*9.* How many africans players are playing in Europe?

*10.* Who is the heaviest player?



In [6]:
df = data[['LongName', 'Nationality', 'Age', '↓OVA', 'Club', 'Joined', 'Contract', 'Height', 'Weight', 'Preferred Foot', 'Best Position','Value', 'Loan Date End', 'Wage']]

In [7]:
df.shape

(18979, 14)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   LongName        18979 non-null  object
 1   Nationality     18979 non-null  object
 2   Age             18979 non-null  int64 
 3   ↓OVA            18979 non-null  int64 
 4   Club            18979 non-null  object
 5   Joined          18979 non-null  object
 6   Contract        18979 non-null  object
 7   Height          18979 non-null  object
 8   Weight          18979 non-null  object
 9   Preferred Foot  18979 non-null  object
 10  Best Position   18979 non-null  object
 11  Value           18979 non-null  object
 12  Loan Date End   1013 non-null   object
 13  Wage            18979 non-null  object
dtypes: int64(2), object(12)
memory usage: 2.0+ MB


## Data Cleaning

In [9]:
df.isnull().sum()

LongName              0
Nationality           0
Age                   0
↓OVA                  0
Club                  0
Joined                0
Contract              0
Height                0
Weight                0
Preferred Foot        0
Best Position         0
Value                 0
Loan Date End     17966
Wage                  0
dtype: int64

#### Longname

In [10]:
# checking for null values
df['LongName'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 18979 entries, 0 to 18978
Series name: LongName
Non-Null Count  Dtype 
--------------  ----- 
18979 non-null  object
dtypes: object(1)
memory usage: 148.4+ KB


In [11]:
#checking for all unique values in the column
df['LongName'].nunique()

18852

In [12]:
#printing out all repeated values in the row
df[df['LongName'].duplicated()]

LongName          Nationality  Age  ↓OVA  \
1239    Lisandro López            Argentina   30    76   
2511        Ben Davies              England   24    73   
2517    Gonzalo Castro              Germany   33    73   
2957     Claudio Bravo            Argentina   23    73   
4275   Ladislav Krejčí       Czech Republic   21    71   
...                ...                  ...  ...   ...   
18707        Peng Wang             China PR   22    51   
18765      Alfie Jones              England   19    50   
18782       Joe Wright              England   19    50   
18822       Mark Byrne  Republic of Ireland   19    50   
18865        Wei Zhang             China PR   20    49   

                                 Club        Joined     Contract Height  \
1239             \n\n\n\nBoca Juniors   Jan 1, 2020  2020 ~ 2023  188cm   
2511        \n\n\n\nPreston North End   Jan 1, 2013  2013 ~ 2021  185cm   
2517            \n\n\n\nVfB Stuttgart   Jul 1, 2018  2018 ~ 2021  172cm   
2957   \n\n\n\nClub Atlético Banfield   Jan 4, 2016  2016 ~ 2021  170cm   
4275             \n\n\n\nSparta Praha   Jul 2, 2019  2019 ~ 2023  190cm   
...                               ...           ...          ...    ...   
18707        \n\n\n\nGuangzhou R&F FC  Aug 10, 2020  2020 ~ 2023  176cm   
18765            \n\n\n\nCrawley Town   Aug 8, 2019  2019 ~ 2021  186cm   
18782                \n\n\n\nMillwall   Jul 1, 2019  2019 ~ 2024  185cm   
18822            \n\n\n\nSligo Rovers  Feb 13, 2020  2020 ~ 2020  178cm   
18865  \n\n\n\nHebei China Fortune FC   Jan 1, 2019  2019 ~ 2024  186cm   

      Weight Preferred Foot Best Position  Value Loan Date End  Wage  
1239    80kg          Right            CB    €6M           NaN  €16K  
2511    74kg           Left            CB  €4.4M           NaN  €18K  
2517    72kg          Right            CM  €1.9M           NaN  €20K  
2957    69kg           Left           LWB  €6.5M           NaN  €10K  
4275    86kg           Left            CB  €3.6M           NaN  €500  
...      ...            ...           ...    ...           ...   ...  
18707   69kg          Right           CAM  €130K           NaN   €2K  
18765   73kg          Right            GK  €110K           NaN  €500  
18782   82kg          Right            GK  €100K           NaN  €650  
18822   74kg          Right           CAM  €110K           NaN  €500  
18865   74kg          Right           CDM  €100K           NaN   €1K  

[127 rows x 14 columns]

In [13]:
#deleting all duplicated values while keeping the original value
df = df.drop_duplicates(subset = ['LongName'])

In [14]:
df['LongName'].head()

0                    Lionel Messi
1    C. Ronaldo dos Santos Aveiro
2                       Jan Oblak
3                 Kevin De Bruyne
4      Neymar da Silva Santos Jr.
Name: LongName, dtype: object

In [15]:
#verifying that there is no duplicated values in the column.
df[df['LongName'].duplicated()]

Empty DataFrame
Columns: [LongName, Nationality, Age, ↓OVA, Club, Joined, Contract, Height, Weight, Preferred Foot, Best Position, Value, Loan Date End, Wage]
Index: []

#### Nationality

In [16]:
df['Nationality'].info()

<class 'pandas.core.series.Series'>
Int64Index: 18852 entries, 0 to 18978
Series name: Nationality
Non-Null Count  Dtype 
--------------  ----- 
18852 non-null  object
dtypes: object(1)
memory usage: 294.6+ KB


In [17]:
df['Nationality'].nunique()

164

In [18]:
df['Nationality'].head()

0    Argentina
1     Portugal
2     Slovenia
3      Belgium
4       Brazil
Name: Nationality, dtype: object

In [19]:
df['Nationality'].shape

(18852,)

In [20]:
df['Nationality'].isnull().sum()

0

*It shows that all the 18852 players under consideration has their own nationality known.*

In [21]:
df.head()

LongName Nationality  Age  ↓OVA  \
0                  Lionel Messi   Argentina   33    93   
1  C. Ronaldo dos Santos Aveiro    Portugal   35    92   
2                     Jan Oblak    Slovenia   27    91   
3               Kevin De Bruyne     Belgium   29    91   
4    Neymar da Silva Santos Jr.      Brazil   28    91   

                          Club        Joined     Contract Height Weight  \
0         \n\n\n\nFC Barcelona   Jul 1, 2004  2004 ~ 2021  170cm   72kg   
1             \n\n\n\nJuventus  Jul 10, 2018  2018 ~ 2022  187cm   83kg   
2      \n\n\n\nAtlético Madrid  Jul 16, 2014  2014 ~ 2023  188cm   87kg   
3      \n\n\n\nManchester City  Aug 30, 2015  2015 ~ 2023  181cm   70kg   
4  \n\n\n\nParis Saint-Germain   Aug 3, 2017  2017 ~ 2022  175cm   68kg   

  Preferred Foot Best Position    Value Loan Date End   Wage  
0           Left            RW  €103.5M           NaN  €560K  
1          Right            ST     €63M           NaN  €220K  
2          Right            GK    €120M           NaN  €125K  
3          Right           CAM    €129M           NaN  €370K  
4          Right            LW    €132M           NaN  €270K

#### Club
> There is a need to remove all the \n in front of the club name for this column.

In [22]:
df['Club'] = df['Club'].map(lambda x: x.lstrip('\n'))

In [23]:
df['Club'].head()

0           FC Barcelona
1               Juventus
2        Atlético Madrid
3        Manchester City
4    Paris Saint-Germain
Name: Club, dtype: object

#### Joined

In [24]:
df['Joined'].info()

<class 'pandas.core.series.Series'>
Int64Index: 18852 entries, 0 to 18978
Series name: Joined
Non-Null Count  Dtype 
--------------  ----- 
18852 non-null  object
dtypes: object(1)
memory usage: 294.6+ KB


This column refers to the date that the players joined the football club they are playing for. 

The column datatype needs to be change to date instead of object.

In [25]:
from datetime import datetime

In [26]:
df['Joined'] = pd.to_datetime(df['Joined'])

In [27]:
df['Joined'].head()

0   2004-07-01
1   2018-07-10
2   2014-07-16
3   2015-08-30
4   2017-08-03
Name: Joined, dtype: datetime64[ns]

#### Contract

In [28]:
df['Contract'].head()

0    2004 ~ 2021
1    2018 ~ 2022
2    2014 ~ 2023
3    2015 ~ 2023
4    2017 ~ 2022
Name: Contract, dtype: object

Considering the format of the *Contract* column, the column will be divided into two column to short the start and the end of the contract then the difference will be calculated to get the exact contract per year. 

In [29]:
df['Contract'].unique()

array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 On Loan',
       '2009 ~ 2021', '2019 ~ 2021', '2019 ~ 2026', 'Free', '2012 ~ 

In [30]:
df[df['Contract'].str.contains('Loan')].index

Int64Index([  205,   248,   254,   302,   306,   332,   336,   405,   416,
              435,
            ...
            18231, 18232, 18241, 18323, 18356, 18472, 18571, 18600, 18622,
            18680],
           dtype='int64', length=1003)

In [31]:
df1 = df[df['Contract'].str.contains('Loan')]

In [32]:
df1['Contract'] = df['Contract'].map(lambda x: x.rstrip(' On Loan'))

C:\Users\Taiwo_Victoria\AppData\Local\Temp\ipykernel_7300\2455802587.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Contract'] = df['Contract'].map(lambda x: x.rstrip(' On Loan'))


In [33]:
df1

LongName Nationality  Age  ↓OVA  \
205                  Gareth Bale       Wales   30    83   
248              Alphonse Areola      France   27    82   
254    Danilo Luís Hélio Pereira    Portugal   28    82   
302              Matteo Politano       Italy   26    81   
306               Lucas Torreira     Uruguay   24    81   
...                          ...         ...  ...   ...   
18472               Moudi Najjar   Australia   20    52   
18571                Arthur Read     England   20    51   
18600                Yubiao Deng    China PR   23    51   
18622                    Pu Chen    China PR   23    51   
18680               Zhechao Chen    China PR   25    51   

                                Club     Joined      Contract Height Weight  \
205                Tottenham Hotspur 2013-09-01  Jun 30, 2021  185cm   82kg   
248                           Fulham 2010-07-01  Jun 30, 2021  195cm   94kg   
254              Paris Saint-Germain 2015-07-02  Jun 30, 2021  188cm   83kg   
302                           Napoli 2019-07-01  Jun 30, 2021  171cm   67kg   
306                  Atlético Madrid 2018-07-12  Jun 30, 2021  166cm   60kg   
...                              ...        ...           ...    ...    ...   
18472                   Macarthur FC 2018-02-13  Aug 31, 2021  182cm   72kg   
18571                      Stevenage 2019-07-12  Jun 30, 2021  178cm   64kg   
18600  Shijiazhuang Ever Bright F.C. 2017-07-14  Dec 31, 2020  181cm   65kg   
18622  Shijiazhuang Ever Bright F.C. 2019-07-25  Dec 31, 2020  177cm   68kg   
18680               Guangzhou R&F FC 2016-01-01  Dec 31, 2020  177cm   70kg   

      Preferred Foot Best Position   Value Loan Date End   Wage  
205             Left            RW    €35M  Jun 30, 2021  €200K  
248            Right            GK    €32M  Jun 30, 2021   €71K  
254            Right           CDM    €27M  Jun 30, 2021   €20K  
302             Left            RM    €28M  Jun 30, 2021  €105K  
306            Right           CDM  €33.5M  Jun 30, 2021   €74K  
...              ...           ...     ...           ...    ...  
18472           Left            ST   €190K  Aug 31, 2021    €2K  
18571           Left           CAM   €140K  Jun 30, 2021    €5K  
18600          Right           CAM   €130K  Dec 31, 2020    €3K  
18622          Right           CAM   €130K  Dec 31, 2020    €2K  
18680           Left            CB    €90K  Dec 31, 2020    €3K  

[1003 rows x 14 columns]

In [34]:
df1['Contract'].equals(df1['Loan Date End']) 

True

Since the contract and loan date end are equal, the loan date end column can be droped and the value of the contract column will be replace with "On Loan" before adding it to the other dataframe.

In [35]:
df1 = df1.drop("Loan Date End", axis='columns')

In [36]:
df1.head()

LongName Nationality  Age  ↓OVA                 Club  \
205                Gareth Bale       Wales   30    83    Tottenham Hotspur   
248            Alphonse Areola      France   27    82               Fulham   
254  Danilo Luís Hélio Pereira    Portugal   28    82  Paris Saint-Germain   
302            Matteo Politano       Italy   26    81               Napoli   
306             Lucas Torreira     Uruguay   24    81      Atlético Madrid   

        Joined      Contract Height Weight Preferred Foot Best Position  \
205 2013-09-01  Jun 30, 2021  185cm   82kg           Left            RW   
248 2010-07-01  Jun 30, 2021  195cm   94kg          Right            GK   
254 2015-07-02  Jun 30, 2021  188cm   83kg          Right           CDM   
302 2019-07-01  Jun 30, 2021  171cm   67kg           Left            RM   
306 2018-07-12  Jun 30, 2021  166cm   60kg          Right           CDM   

      Value   Wage  
205    €35M  €200K  
248    €32M   €71K  
254    €27M   €20K  
302    €28M  €105K  
306  €33.5M   €74K

In [37]:
df2 = df[df['Contract'].str.contains('~')]

In [38]:
df2.head()

LongName Nationality  Age  ↓OVA                 Club  \
0                  Lionel Messi   Argentina   33    93         FC Barcelona   
1  C. Ronaldo dos Santos Aveiro    Portugal   35    92             Juventus   
2                     Jan Oblak    Slovenia   27    91      Atlético Madrid   
3               Kevin De Bruyne     Belgium   29    91      Manchester City   
4    Neymar da Silva Santos Jr.      Brazil   28    91  Paris Saint-Germain   

      Joined     Contract Height Weight Preferred Foot Best Position    Value  \
0 2004-07-01  2004 ~ 2021  170cm   72kg           Left            RW  €103.5M   
1 2018-07-10  2018 ~ 2022  187cm   83kg          Right            ST     €63M   
2 2014-07-16  2014 ~ 2023  188cm   87kg          Right            GK    €120M   
3 2015-08-30  2015 ~ 2023  181cm   70kg          Right           CAM    €129M   
4 2017-08-03  2017 ~ 2022  175cm   68kg          Right            LW    €132M   

  Loan Date End   Wage  
0           NaN  €560K  
1           NaN  €220K  
2           NaN  €125K  
3           NaN  €370K  
4           NaN  €270K

In [39]:
# dividing the column into to with a delimeter ( ~ )
df2[['Start','End']] = df2.Contract.str.split(" ~ ",expand=True)

C:\Users\Taiwo_Victoria\AppData\Local\Temp\ipykernel_7300\2220130809.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2[['Start','End']] = df2.Contract.str.split(" ~ ",expand=True)
C:\Users\Taiwo_Victoria\AppData\Local\Temp\ipykernel_7300\2220130809.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2[['Start','End']] = df2.Contract.str.split(" ~ ",expand=True)


In [40]:
# converting the *Start* and *End* column to int

df2['Start'] = pd.to_numeric(df2['Start'])
df2['End'] = pd.to_numeric(df2['End'])

C:\Users\Taiwo_Victoria\AppData\Local\Temp\ipykernel_7300\1441081867.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['Start'] = pd.to_numeric(df2['Start'])
C:\Users\Taiwo_Victoria\AppData\Local\Temp\ipykernel_7300\1441081867.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['End'] = pd.to_numeric(df2['End'])


In [41]:
df2['Contract_Period'] = df2['End'] - df2['Start']

C:\Users\Taiwo_Victoria\AppData\Local\Temp\ipykernel_7300\3021308082.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['Contract_Period'] = df2['End'] - df2['Start']


In [42]:
df2['Contract_Period'].info()

<class 'pandas.core.series.Series'>
Int64Index: 17612 entries, 0 to 18978
Series name: Contract_Period
Non-Null Count  Dtype
--------------  -----
17612 non-null  int64
dtypes: int64(1)
memory usage: 275.2 KB


In [43]:
df2['Contract_Period'].unique()

array([17,  4,  9,  8,  5,  6,  7, 10, 12, 13, 16, 14, 15,  2, 11,  3,  1,
       20,  0, 18, 21, 19, 23], dtype=int64)

In [44]:
df2 = df2.drop("Loan Date End", axis='columns')

In [45]:
df2.head()

LongName Nationality  Age  ↓OVA                 Club  \
0                  Lionel Messi   Argentina   33    93         FC Barcelona   
1  C. Ronaldo dos Santos Aveiro    Portugal   35    92             Juventus   
2                     Jan Oblak    Slovenia   27    91      Atlético Madrid   
3               Kevin De Bruyne     Belgium   29    91      Manchester City   
4    Neymar da Silva Santos Jr.      Brazil   28    91  Paris Saint-Germain   

      Joined     Contract Height Weight Preferred Foot Best Position    Value  \
0 2004-07-01  2004 ~ 2021  170cm   72kg           Left            RW  €103.5M   
1 2018-07-10  2018 ~ 2022  187cm   83kg          Right            ST     €63M   
2 2014-07-16  2014 ~ 2023  188cm   87kg          Right            GK    €120M   
3 2015-08-30  2015 ~ 2023  181cm   70kg          Right           CAM    €129M   
4 2017-08-03  2017 ~ 2022  175cm   68kg          Right            LW    €132M   

    Wage  Start   End  Contract_Period  
0  €560K   2004  2021               17  
1  €220K   2018  2022                4  
2  €125K   2014  2023                9  
3  €370K   2015  2023                8  
4  €270K   2017  2022                5

In [46]:
On_Contract = df2.drop("Contract", axis='columns')

In [47]:
On_Contract.head()

LongName Nationality  Age  ↓OVA                 Club  \
0                  Lionel Messi   Argentina   33    93         FC Barcelona   
1  C. Ronaldo dos Santos Aveiro    Portugal   35    92             Juventus   
2                     Jan Oblak    Slovenia   27    91      Atlético Madrid   
3               Kevin De Bruyne     Belgium   29    91      Manchester City   
4    Neymar da Silva Santos Jr.      Brazil   28    91  Paris Saint-Germain   

      Joined Height Weight Preferred Foot Best Position    Value   Wage  \
0 2004-07-01  170cm   72kg           Left            RW  €103.5M  €560K   
1 2018-07-10  187cm   83kg          Right            ST     €63M  €220K   
2 2014-07-16  188cm   87kg          Right            GK    €120M  €125K   
3 2015-08-30  181cm   70kg          Right           CAM    €129M  €370K   
4 2017-08-03  175cm   68kg          Right            LW    €132M  €270K   

   Start   End  Contract_Period  
0   2004  2021               17  
1   2018  2022                4  
2   2014  2023                9  
3   2015  2023                8  
4   2017  2022                5

In [48]:
On_Contract = On_Contract.drop("Start", axis='columns')
On_Contract = On_Contract.drop("End", axis='columns')

In [49]:
On_Contract.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17612 entries, 0 to 18978
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   LongName         17612 non-null  object        
 1   Nationality      17612 non-null  object        
 2   Age              17612 non-null  int64         
 3   ↓OVA             17612 non-null  int64         
 4   Club             17612 non-null  object        
 5   Joined           17612 non-null  datetime64[ns]
 6   Height           17612 non-null  object        
 7   Weight           17612 non-null  object        
 8   Preferred Foot   17612 non-null  object        
 9   Best Position    17612 non-null  object        
 10  Value            17612 non-null  object        
 11  Wage             17612 non-null  object        
 12  Contract_Period  17612 non-null  int64         
dtypes: datetime64[ns](1), int64(3), object(9)
memory usage: 1.9+ MB


In [50]:
On_Loan = df1

On_Loan.head()

LongName Nationality  Age  ↓OVA                 Club  \
205                Gareth Bale       Wales   30    83    Tottenham Hotspur   
248            Alphonse Areola      France   27    82               Fulham   
254  Danilo Luís Hélio Pereira    Portugal   28    82  Paris Saint-Germain   
302            Matteo Politano       Italy   26    81               Napoli   
306             Lucas Torreira     Uruguay   24    81      Atlético Madrid   

        Joined      Contract Height Weight Preferred Foot Best Position  \
205 2013-09-01  Jun 30, 2021  185cm   82kg           Left            RW   
248 2010-07-01  Jun 30, 2021  195cm   94kg          Right            GK   
254 2015-07-02  Jun 30, 2021  188cm   83kg          Right           CDM   
302 2019-07-01  Jun 30, 2021  171cm   67kg           Left            RM   
306 2018-07-12  Jun 30, 2021  166cm   60kg          Right           CDM   

      Value   Wage  
205    €35M  €200K  
248    €32M   €71K  
254    €27M   €20K  
302    €28M  €105K  
306  €33.5M   €74K

In [51]:
On_Loan.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1003 entries, 205 to 18680
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   LongName        1003 non-null   object        
 1   Nationality     1003 non-null   object        
 2   Age             1003 non-null   int64         
 3   ↓OVA            1003 non-null   int64         
 4   Club            1003 non-null   object        
 5   Joined          1003 non-null   datetime64[ns]
 6   Contract        1003 non-null   object        
 7   Height          1003 non-null   object        
 8   Weight          1003 non-null   object        
 9   Preferred Foot  1003 non-null   object        
 10  Best Position   1003 non-null   object        
 11  Value           1003 non-null   object        
 12  Wage            1003 non-null   object        
dtypes: datetime64[ns](1), int64(2), object(10)
memory usage: 109.7+ KB


In [52]:
On_Loan['Contract'] = pd.to_datetime(On_Loan['Contract'])

In [58]:
On_Loan['Loan_Period'] = On_Loan['Contract'] - On_Loan['Joined']

In [55]:
On_Loan.head()

LongName Nationality  Age  ↓OVA                 Club  \
205                Gareth Bale       Wales   30    83    Tottenham Hotspur   
248            Alphonse Areola      France   27    82               Fulham   
254  Danilo Luís Hélio Pereira    Portugal   28    82  Paris Saint-Germain   
302            Matteo Politano       Italy   26    81               Napoli   
306             Lucas Torreira     Uruguay   24    81      Atlético Madrid   

        Joined   Contract Height Weight Preferred Foot Best Position   Value  \
205 2013-09-01 2021-06-30  185cm   82kg           Left            RW    €35M   
248 2010-07-01 2021-06-30  195cm   94kg          Right            GK    €32M   
254 2015-07-02 2021-06-30  188cm   83kg          Right           CDM    €27M   
302 2019-07-01 2021-06-30  171cm   67kg           Left            RM    €28M   
306 2018-07-12 2021-06-30  166cm   60kg          Right           CDM  €33.5M   

      Wage Loan_Period  
205  €200K   2859 days  
248   €71K   4017 days  
254   €20K   2190 days  
302  €105K    730 days  
306   €74K   1084 days

In [59]:
df.head()

LongName Nationality  Age  ↓OVA                 Club  \
0                  Lionel Messi   Argentina   33    93         FC Barcelona   
1  C. Ronaldo dos Santos Aveiro    Portugal   35    92             Juventus   
2                     Jan Oblak    Slovenia   27    91      Atlético Madrid   
3               Kevin De Bruyne     Belgium   29    91      Manchester City   
4    Neymar da Silva Santos Jr.      Brazil   28    91  Paris Saint-Germain   

      Joined     Contract Height Weight Preferred Foot Best Position    Value  \
0 2004-07-01  2004 ~ 2021  170cm   72kg           Left            RW  €103.5M   
1 2018-07-10  2018 ~ 2022  187cm   83kg          Right            ST     €63M   
2 2014-07-16  2014 ~ 2023  188cm   87kg          Right            GK    €120M   
3 2015-08-30  2015 ~ 2023  181cm   70kg          Right           CAM    €129M   
4 2017-08-03  2017 ~ 2022  175cm   68kg          Right            LW    €132M   

  Loan Date End   Wage  
0           NaN  €560K  
1           NaN  €220K  
2           NaN  €125K  
3           NaN  €370K  
4           NaN  €270K

In [60]:
df = df.drop(["Loan Date End", "Contract"], axis='columns')

In [61]:
df.head()

LongName Nationality  Age  ↓OVA                 Club  \
0                  Lionel Messi   Argentina   33    93         FC Barcelona   
1  C. Ronaldo dos Santos Aveiro    Portugal   35    92             Juventus   
2                     Jan Oblak    Slovenia   27    91      Atlético Madrid   
3               Kevin De Bruyne     Belgium   29    91      Manchester City   
4    Neymar da Silva Santos Jr.      Brazil   28    91  Paris Saint-Germain   

      Joined Height Weight Preferred Foot Best Position    Value   Wage  
0 2004-07-01  170cm   72kg           Left            RW  €103.5M  €560K  
1 2018-07-10  187cm   83kg          Right            ST     €63M  €220K  
2 2014-07-16  188cm   87kg          Right            GK    €120M  €125K  
3 2015-08-30  181cm   70kg          Right           CAM    €129M  €370K  
4 2017-08-03  175cm   68kg          Right            LW    €132M  €270K

In [77]:
from datetime import date

today = pd.to_datetime(date.today())

In [92]:
df['diff'] = (today - df['Joined'])  // 365

In [94]:
df['diff']

0       18 days 14:35:50.136986301
1        4 days 13:48:29.589041095
2        8 days 13:28:46.027397260
3        7 days 10:31:13.972602739
4        5 days 12:13:48.493150684
                   ...            
18974    4 days 13:36:39.452054794
18975    2 days 12:17:45.205479452
18976    3 days 21:57:41.917808219
18977    2 days 08:52:36.164383561
18978    3 days 12:33:32.054794520
Name: diff, Length: 18852, dtype: timedelta64[ns]

In [63]:
df.head()

LongName Nationality  Age  ↓OVA                 Club  \
0                  Lionel Messi   Argentina   33    93         FC Barcelona   
1  C. Ronaldo dos Santos Aveiro    Portugal   35    92             Juventus   
2                     Jan Oblak    Slovenia   27    91      Atlético Madrid   
3               Kevin De Bruyne     Belgium   29    91      Manchester City   
4    Neymar da Silva Santos Jr.      Brazil   28    91  Paris Saint-Germain   

      Joined Height Weight Preferred Foot Best Position    Value   Wage  \
0 2004-07-01  170cm   72kg           Left            RW  €103.5M  €560K   
1 2018-07-10  187cm   83kg          Right            ST     €63M  €220K   
2 2014-07-16  188cm   87kg          Right            GK    €120M  €125K   
3 2015-08-30  181cm   70kg          Right           CAM    €129M  €370K   
4 2017-08-03  175cm   68kg          Right            LW    €132M  €270K   

   Joined_period  
0             18  
1              4  
2              8  
3              7  
4              5